### PR TITLE
feature(stress-thread): set timeout based on duration

### PR DESCRIPTION
**Pull request by soyacz**
Sometimes we face freeze of c-s thread blocking the test flow.

This commit adds automatic timeout to stress threads (c-s, sb and
gemini) when duration in stress cmd is specified. 
Timeout is calculated as this:
warmup + duration + 600 (all in seconds).

https://trello.com/c/wZoe5pd3

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sdcm/utils/functions.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# This program is free software; you can redistribute it and/or modify
```
Comment (fruch) [inline sdcm/utils/functions.py:1]: Module named function ? those are function to parse stress commands

Why not put those functions on the stress classes ? so each thread could compute the duration/timeout ?

**Review by fruch — APPROVED** [review]
LGTM

File: sdcm/stress_thread.py [hunk 2]
```diff
@@ -370,3 +370,27 @@ def db_node_to_query(self, loader):
             LOGGER.debug("Selected '%s' to query for local nodes", node_to_query)
             return node_to_query.cql_ip_address
         return self.node_list[0].cql_ip_address
+
+
+duration_pattern = re.compile(r'(?P<hours>[\d]*)h|(?P<minutes>[\d]*)m|(?P<seconds>[\d]*)s')
+stress_cmd_get_duration_pattern = re.compile(r' [-]{0,2}duration[\s=]+([\d]+[hms]+)')
+stress_cmd_get_warmup_pattern = re.compile(r' [-]{0,2}warmup[\s=]+([\d]+[hms]+)')
```
Comment (fgelcer) [inline sdcm/stress_thread.py:377]: problem here is that usually warmup/prepare workloads don't have a duration, but an amount of writes predefined...
in case it is not found, we will most likely miss the whole prepare time in the test duration

File: sdcm/stress_thread.py [hunk 3]
```diff
@@ -370,3 +370,27 @@ def db_node_to_query(self, loader):
...
+
+
+def time_period_str_to_seconds(time_str: str) -> int:
+    """Transforms duration string into seconds int. e.g. 1h -> 3600, 1h22m->4920 or 10m->600"""
+    return sum([int(g[0] or 0) * 3600 + int(g[1] or 0) * 60 + int(g[2] or 0) for g in duration_pattern.findall(time_str)])
```
Comment (fgelcer) [inline sdcm/stress_thread.py:382]: what is `g`?
couldn't it have a better name?

**Review by fgelcer — COMMENTED** [review]
here an example of prepare command (`longevity-100gb-4h.yaml`)

```
prepare_write_cmd:  "cassandra-stress write cl=QUORUM n=20971520 -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -mode cql3 native -rate threads=80 -pop seq=1..20971520 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
```

File: sdcm/stress_thread.py (hunk 2 above)
Comment (fruch) [reply sdcm/stress_thread.py:377]: So it would default to the test duration, as it was before this PR, what's the issue ?

Comment (fgelcer) [reply sdcm/stress_thread.py:377]: > So it would default to the test duration, as it was before this PR, what's the issue ?

no, this is not what is happening in here...
in the code below (that was taken from `sdcm/stress_thread.py` proposed in this PR) we will attempt to get the time duration for each command, and if one doesn't have any entry (but at least one has) we will continue

```
def get_timeout_from_stress_cmd(stress_cmd: str) -> int | None:
    """Gets timeout in seconds based on duration and warmup arguments from stress command."""
    timeout = 0
    if duration_match := stress_cmd_get_duration_pattern.search(stress_cmd):
        timeout += time_period_str_to_seconds(duration_match.group(0))
    if warmup_match := stress_cmd_get_warmup_pattern.search(stress_cmd):
        timeout += time_period_str_to_seconds(warmup_match.group(0))
    if timeout == 0:
        return None
    else:
        # adding 10 minutes to timeout for general all others delays
        return timeout + 600
```

also, there might be another problem with it, that if we have split a command into many (like few writes to use round robin) we will sum up all the values... or i'm misunderstanding the code in here?

Comment (fruch) [reply sdcm/stress_thread.py:377]: You are missing the part where the timeout was taken from before this PR

Comment (fgelcer) [reply sdcm/stress_thread.py:377]: it was taken from the `test_duration` param in the test's `.yaml` file

Comment (fruch) [reply sdcm/stress_thread.py:377]: and it's still the case if the time isn't extracted from the stress command, this function would return None, and the default would be the test duration

Comment (fgelcer) [reply sdcm/stress_thread.py:377]: let me rephrase

if we have duration of the stress command, but we don't have for `prepare` or `warmup` we will not get `None`, but only the value for the stress command, and in this case, if the prepare takes long, the test might timeout before the `timeout` is completed, because it will be lacking part of the stresses it needs to run...

let's look at this example from `longevity-100gb-4h.yaml`:
```
prepare_write_cmd:  "cassandra-stress write cl=QUORUM n=20971520 -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -mode cql3 native -rate threads=80 -pop seq=1..20971520 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"

stress_cmd: ["cassandra-stress write cl=QUORUM duration=240m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -mode cql3 native -rate threads=40 -pop seq=1..20971520 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5",
             "cassandra-stress read  cl=QUORUM duration=240m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -mode cql3 native -rate threads=40 -pop seq=1..20971520 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
             ]
```

the `prepare_write_cmd` does not have a timeout defined in it, so it wouldn't be accounted into the `duration=240m`

so if the prepare takes more than the 10 min buffer set in the following line `return timeout + 600` the test will "timeout" before completed..

also, in case we have 2 `stress_cmd` like in this example... would it be accounted as `duration=240m` + `duration=240m` ?


I understand that in case we didn't find any of duration we will continue to do what is being done before this PR... or am I missing something else?

Comment (soyacz) [reply sdcm/stress_thread.py:377]: c-s is not having warmup duration parameter (at least I didn't see it). Only `no-warmup`. For this (not only) we add this `600` seconds.  Duration param is present in scylla bench.
If there's no duration specified we fallback to current solution = test duration.
In case we have multiple stress_cmd's, then anyway each c-s is executed in separate thread and each gets timeout according to duration.
In case of doubts, see unit tests, also propose one if current don't cover some case.

File: sdcm/stress_thread.py (hunk 3 above)
Comment (soyacz) [reply sdcm/stress_thread.py:382]: possibly it could have better name,`g` is just for brevity (it is from `group`). Function docs should be enough to understand how it works.

File: sdcm/stress_thread.py (hunk 2 above)
Comment (fgelcer) [reply sdcm/stress_thread.py:377]: ah, this will be for the thread only... ok, i see it now...

**Review by fgelcer — APPROVED** [review]
LGTM
